include/ppx/knob.h

```diff
@@ -0,0 +1,284 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ppx_knob_h
+#define ppx_knob_h
+
+#include "ppx/command_line_parser.h"
+#include "ppx/config.h"
+#include "ppx/log.h"
+
+#include <functional>
+#include <iostream>
+#include <unordered_set>
+#include <vector>
+
+namespace ppx {
+
+// ---------------------------------------------------------------------------------------------
+// Knob Classes
+// ---------------------------------------------------------------------------------------------
+
+// Knob is an abstract class which contains common features for all knobs and knob hierarchy
+class Knob
```

Comment:
Taking a step back: why do knobs have children?

Is it only for UI grouping (like ImGUI groups), or is it something else?
What if a slider has N children? Does the children behavior depends on this node?

Maybe the children part shouldn't be part of the Knob itself, but something else.
Like ImGUI groups, which are not imGUI interactive widgets.

Maybe the "Manager" could be renamed "Group", and just have a linear list of knobs to display, while knobs don't have this hierarchy?

Reply:
It seems it's UI grouping only.

KnobManager should be owning the knobs, so their lifetime management is not going to be a mess.

Reply:
Well, since we have UI groupd, do we need a manager then?
1 class method is the root-level group, which has N children. Only the KnobGroup has children, as then, there is no point for sliders/buttons to have groups.

Having another class which does 1 think which is to store children referenced by others doesn't seem to simplify the lifetime management. The lifetime of the root would determine the lifecycle of the children no?

Reply:
> I originally wanted knobs to have children so that there can be some sort of visual organization in the GUI

Yes, understand this point. But for the checkbox example: how are children of the checkbox impacted? Like if the group folded when the checkbox is shown, or not? If yes, how does it work for a slider, as the value is not a boolean anymore.

ImGui decision is quite straightforward: components have no children, except for Groups, which don't contains interactive elements, but organize children UI components.
In your case, what about not having children for KnobSlider by ex (and thus the parent Knob class), and only have children on the KnobGroup item?

> I'm not sure that I understand the alternatives proposed here for not having a manager.

Right now, knobs know how to draw themselves, and they iterate through their children, and call child.draw(). So maybe the manager is not needed as you could have 1 root knob, which is a group, with N knobs as children.

An example can be seen on this gist (https://gist.github.com/Keenuts/cc0853887683797d5ba158bb8b11c65c).
In this sample, you have no manager, as each knob parent manages/own their children.
This way, you don't have the same knob referenced by a manager and the parent. Ownership is straightforward.

Reply:
We can always add grouping later (if we ever needs it)